### PR TITLE
=osgi #19323 Add to documentation on OSGi

### PR DESCRIPTION
**Pull request by briantopping**
Carryover from #19364, sorry about that, pooched the branch.

### CODE REVIEW

File: akka-docs/rst/additional/osgi.rst [hunk 1]
```diff
@@ -1,12 +1,97 @@
...
+--------------------------------------------------
+
+The fundamental unit of deployment in OSGi is the ``Bundle``. A bundle is a Java JAR with `additional
+entries <https://www.osgi.org/bundle-headers-reference/>` in ``MANIFEST.MF`` that minimally expose the name and version
+of the bundle and packages for import and export. Since these manifest entries are ignored outside OSGi deployments,
+a bundle can interchangeably be used as a JAR in the JRE.
+
+When a bundle is loaded, a specialized implementation of the Java ``ClassLoader`` is instantiated for each bundle. Each
+classloader reads the manifest entries and publishes both it's capabilities (in the form of the bundle's exports) and
+requirements (it's imports) in a container singleton for discovery by other bundles. The process of matching imports to
```
Comment (patriknw) [inline akka-docs/rst/additional/osgi.rst]: s/it's/its/

File: akka-docs/rst/additional/osgi.rst [hunk 2]
```diff
@@ -1,12 +1,97 @@
...
+exports across bundles through these classloaders is the process of resolution, one of six discrete steps in the lifecycle
+FSM of a bundle in an OSGi container:
+
+1. INSTALLED: A bundle that is installed has been loaded from disk and a classloader instantiated with it's capabilities.
```
Comment (patriknw) [inline akka-docs/rst/additional/osgi.rst]: its

File: akka-docs/rst/additional/osgi.rst [hunk 3]
```diff
@@ -1,12 +1,97 @@
...
+   Bundles are iteratively installed manually or through container-specific descriptors. In contrast to frameworks such as
+   EJB or PAR, the modular nature of OSGi means that bundles may be used by multiple applications with overlapping dependencies
```
Comment (patriknw) [inline akka-docs/rst/additional/osgi.rst]: what does EJB has to do with this? what is PAR?

File: akka-docs/rst/additional/osgi.rst [hunk 4]
```diff
@@ -1,12 +1,97 @@
...
+Combined with understanding the bundle lifecycle, the OSGi developer must pay attention to sometimes unexpected behaviors
+that are introduced. These are generally within the JVM specification, but are unexpected and can lead to frustration.
+
+* Bundles should not export overlapping package spaces. It is not uncommon for legacy JVM frameworks to expect plugins
+  in an application composed of multiple JARs to reside under a single package name. For example, a frontend application
+  might scan all classes from ``com.example.plugins`` for specific service implementations with that package existing in
+  several contributed JARs.
+
+  While it is possible to support overlapping packages with complex manifest headers, it's much better to use non-overlapping
+  package spaces and facilities such as `Akka Cluster <@github@/akka-docs/rst/scala/code/docs/akka/current/common/cluster.html>`_
```
Comment (patriknw) [inline akka-docs/rst/additional/osgi.rst]: this should be an rst link instead

**Comment by patriknw** [comment]
I like this. I, for one, needed that introduction refresher.

**Comment by drewhk** [comment]
I am happy to merge this after minor suggestions were corrected!

**Comment by briantopping** [comment]
Thanks @patriknw for the edits and @drewhk for the note. I wanted to wait a few days to see if there were any other changes and do them at the same time. 

Please take a look and see if the changes cover concerns! I think there's more that can be added over time, but this is a great start.

**Comment by drewhk** [comment]
I think it is fine for now.

File: akka-docs/rst/additional/osgi.rst [hunk 5]
```diff
@@ -1,12 +1,104 @@
...
+  in an application composed of multiple JARs to reside under a single package name. For example, a frontend application
+  might scan all classes from ``com.example.plugins`` for specific service implementations with that package existing in
+  several contributed JARs.
+
+  While it is possible to support overlapping packages with complex manifest headers, it's much better to use non-overlapping
+  package spaces and facilities such as `Akka Cluster`_
+  for service discovery. Stylistically, many organizations opt to use the root package path as the name of the bundle
+  distribution file.
+
+.. _Akka Cluster: @github@/akka-docs/rst/scala/code/docs/akka/current/common/cluster.html
```
Comment (ktoso) [inline akka-docs/rst/additional/osgi.rst]: Is this link valid? We don't have such file in the repo. If you mean to link to docs then it should link via :refs: instead, shouldn't it?

Comment (patriknw) [inline akka-docs/rst/additional/osgi.rst]: It should be an ordinary rst link

Comment (drewhk) [inline akka-docs/rst/additional/osgi.rst]: wow, I missed this.

Comment (ktoso) [inline akka-docs/rst/additional/osgi.rst]: Would you be able to fix that @briantopping? Thanks in advance if yes :)

**Pull request merged**